src/proto/uprotocol/uuri.rs

```diff
@@ -293,19 +300,240 @@ impl TryFrom<&str> for UUri {
+        }
+
+        // UENTITY_VERSION
+        let version = uri
+            .entity
+            .as_ref()
+            .and_then(|entity| entity.version_major)
+            .unwrap_or(0);
+        buf.put_u8(version as u8);
+
+        // UNUSED
+        buf.put_u8(0);
+
+        // UAUTHORITY
+        if let Some(authority) = uri.authority.as_ref() {
+            buf.put(Vec::try_from(authority)?.as_slice());
+        }
+
+        Ok(buf)
+    }
+}
+
+impl TryFrom<&[u8]> for UUri {
+    type Error = SerializationError;
+
+    /// Creates a `UUri` data object from a uProtocol micro URI.
+    ///
+    /// # Arguments
+    ///
+    /// * `micro_uri` - A byte slice representing the uProtocol micro URI.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `UUri` data object.
+    fn try_from(micro_uri: &[u8]) -> Result<Self, Self::Error> {
+        if micro_uri.len() < LOCAL_MICRO_URI_LENGTH {
+            return Err(SerializationError::new("URI is empty or not in micro form"));
+        }
+
+        let mut buf: &[u8] = micro_uri;
+        // Need to be version 1
+        if buf.get_u8() != UP_VERSION {
+            return Err(SerializationError::new(format!(
+                "URI is not of expected uProtocol version {}",
+                UP_VERSION
+            )));
+        }
+        let address_type = AddressType::try_from(buf.get_u8())?;
+
+        match address_type {
+            AddressType::Local => {
+                if micro_uri.len() != LOCAL_MICRO_URI_LENGTH {
+                    return Err(SerializationError::new("Invalid micro URI length"));
+                }
+            }
+            AddressType::IPv4 => {
+                if micro_uri.len() != IPV4_MICRO_URI_LENGTH {
+                    return Err(SerializationError::new("Invalid micro URI length"));
+                }
+            }
+            AddressType::IPv6 => {
+                if micro_uri.len() != IPV6_MICRO_URI_LENGTH {
+                    return Err(SerializationError::new("Invalid micro URI length"));
+                }
+            }
+            AddressType::ID => {
+                // we cannot perform any reasonable check at this point because we do not
+                // know the (variable) length of the authority ID yet
+            }
+        }
+
+        // RESOURCE
+        let uresource_id = u32::from(buf.get_u16());
+        let resource = Some(UResourceBuilder::from_id(uresource_id));
+
+        // UENTITY
+        let ue_id = buf.get_u16();
+        let ue_version = u32::from(buf.get_u8());
+        let entity = Some(UEntity {
+            id: Some(ue_id.into()),
+            version_major: Some(ue_version),
+            ..Default::default()
+        });
+
+        // skip unused byte
+        buf.advance(1);
+
+        // Calculate uAuthority
+        let authority = match address_type {
+            AddressType::IPv4 => Some(UAuthority {
+                number: Some(crate::uprotocol::uri::uauthority::Number::Ip(
```

Comment:
How about simplifying the type by using `use`?

Reply:
Done